doc/doxygen/src/build-system-basics.md

```diff
@@ -0,0 +1,85 @@
+# Build System Basics                                       {#build-system-basics}
+
+# Variables declaration guidelines              {#variable-declaration-guidelines}
+
+These page contains basic guidelines about `make` variable declaration, it
+summarizes some of the pros/cons as well as specifying good and bad patterns
+in our build system. It is highly recommended to also read the official `gnu make`
+documentation regarding these subjects.
+
+## Avoid unnecessary export
+
+```
+export OUTPUT = $(shell some-command)
+```
+
+Exporting a variable means it will be evaluated on every `target` call which slows
+down the build system. Always avoid exporting a variable if unneeded. 
+
+If an export is actually needed by a `sub-make` then export the variable only for
+the needed targets using `target-export-variables` (more in
+`makefiles/utils/variables.mk`)
+
+Exported variables ("global variable") are hard to remove, specially when badly
+documented. If no one knows why its there and no one knows where it can be used
+then no one knows if its safe to remove since its present for every target.
+
+[gnumake doc](https://www.gnu.org/software/make/manual/html_node/Variables_002fRecursion.html)
+
+## Use memoized for variables referencing a function/command
+
+### recursively expanded variable:
+
+~~~~~~~~~~~~~~~~
+OUTPUT = $(shell some-command $(ANOTHER_VARIABLE))
+~~~~~~~~~~~~~~~~
+
+- When using `=` the value of the variable is only declared, but not set,
+  therefore will only be evaluated when expanded (used) somewhere in the
+  makefile. Then if `$(OUTPUT)` is never expanded, `some-command`
+  is never executed and `ANOTHER_VARIABLE` not expanded.
+
+- They will be evaluated every time the variable is expanded. In the example
+  `some-command` is executed every time `OUTPUT` is expanded, same for
+  `ANOTHER_VARIABLE`. If `some-command` is slow this introduced unneeded overhead.
+
+- If the variable expansion doesn't involve evaluating a function the overhead
+  is none.
+
+### simply expanded variable:
+
+~~~~~~~~~~~~~~~~
+OUTPUT := $(shell some-command $(ANOTHER_VARIABLE))
+~~~~~~~~~~~~~~~~
+
+- When using `:=` the value is only expanded once, expanding any reference to
+  other variables or functions. If `OUTPUT` is always used at least once and
+  evaluates a costly function (`some command`) then use `:=`.
+
+- When using `:=` the variable will be evaluated even if not needed, which is not
+  great if `some command` or functions evaluated by `ANOTHER_VARIABLE` are slow.
+  It can also cause a failure in a worse case scneario (think what happens if a
+  tool is defined with `:=` but you don't have the tool and you don't need it either)
+
+- The value of variables defined with `:=` depend on the order they are defined.
```

Comment:
s/value/values/
s/defined with/defined by/

maybe: depend on the order of definition.